server/routes/users.js

```diff
@@ -1,11 +1,88 @@
 const express = require('express');
 const router = express.Router();
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const config = require('config');
+const {
+	body,
+	validationResult,
+} = require('express-validator');
+
+const User = require('../models/User');
 
 // @route   POST api/users
 // @desc    Register a user
 // @access  Public
-router.post('/', (req, res) => {
-	res.send('Register a user');
-});
+router.post(
+	'/',
+	[
+		body('firstname', 'Please include your first name')
+			.not()
+			.isEmpty(),
+		body('lastname', 'Please include your last name')
+			.not()
+			.isEmpty(),
+		body('email', 'Please include a valid email'),
+		body(
+			'password',
+			'Please include a password with 6 or more characters'
+		).isLength({ min: 6 }),
+	],
+	async (req, res) => {
+		const errors = validationResult(req);
+		if (!errors.isEmpty()) {
+			return res
+				.status(400)
+				.json({ errors: errors.array() });
+		}
+
+		const { firstname, lastname, email, password } =
+			req.body;
+
+		try {
+			let user = await User.findOne({ email: email });
+
+			if (user) {
+				return res
+					.status(400)
+					.json({ msg: 'Email already in use' });
+			}
+
+			user = new User({
+				firstname,
+				lastname,
+				email,
+				password,
+			});
+
+			const salt = await bcrypt.genSalt(10);
+
+			user.password = await bcrypt.hash(password, salt);
```

Comment:
Pasting our Discord conversation here for austerity. :)

1. Prevents rainbow table attacks where hashes of common passwords are compared with hashed values stored in our database.
2. Because we won't be decrypting the hashed value, we just use the stored hashed value as-is for comparison only.
3. The value bcrypt gives us to save in our database contains a hash of both the salt and the password but also contains a hash of just the salt. So upon login it takes the user input password and hashes it, takes the hashed salt from the db. Hashes those together. Compares that value with the part of the value in the db which represents the hash of both salt and password.